src/main/java/org/apache/sling/discovery/oak/SlingIdCleanupTask.java

```diff
@@ -0,0 +1,571 @@
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.commons.scheduler.ScheduleOptions;
+import org.apache.sling.commons.scheduler.Scheduler;
+import org.apache.sling.discovery.InstanceDescription;
+import org.apache.sling.discovery.TopologyEvent;
+import org.apache.sling.discovery.TopologyEvent.Type;
+import org.apache.sling.discovery.TopologyEventListener;
+import org.apache.sling.discovery.TopologyView;
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.metatype.annotations.AttributeDefinition;
+import org.osgi.service.metatype.annotations.Designate;
+import org.osgi.service.metatype.annotations.ObjectClassDefinition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A background task that cleans up garbage slingIds after topology changes.
+ * <p>
+ * A slingId is considered garbage when:
+ * <ul>
+ * <li>it is not in the current topology</li>
+ * <li>it is not in the current idmap (where clusterNodeIds are reused hence
+ * that list stays small and does not need cleanup)</li>
+ * <li>its leaderElectionId was created more than 7 days ago (the
+ * leaderElectionId is created at activate time of the discovery.oak bundle -
+ * hence this more or less corresponds to the startup time of that
+ * instance)</li>
+ * </ul>
+ * The garbage accumulates at the following places, where it will thus be
+ * cleaned up from:
+ * <ul>
+ * <li>as child node under /var/discovery/oak/clusterInstances : this is the
+ * most performance critical garbage</li>
+ * <li>as a property key in /var/discovery/oak/syncTokens</li>
+ * </ul>
+ * The task by default is executed:
+ * <ul>
+ * <li>only on the leader</li>
+ * <li>10min after a TOPOLOGY_INIT or TOPOLOGY_CHANGED event</li>
+ * <li>with a maximum number of delete operations to avoid repository overload -
+ * that maximum is called batchSize and is 50 by default</li>
+ * <li>in subsequent intervals of 10min after the initial run, if that had to
+ * stop at the batchSize of 50 deletions</li>
+ * </ul>
+ * All parameters mentioned above can be configured.
+ */
+@Component
+@Designate(ocd = SlingIdCleanupTask.Conf.class)
+public class SlingIdCleanupTask implements TopologyEventListener, Runnable {
+
+    final static String SLINGID_CLEANUP_ENABLED_SYSTEM_PROPERTY_NAME = "org.apache.sling.discovery.oak.slingidcleanup.enabled";
+
+    /**
+     * default age is 1 week : an instance that is not in the current topology,
+     * started 1 week ago is very unlikely to still be active
```

Comment:
yes that's a bit the thing : an instance can also be long living and not change slingIds - depends on the how this is operated. So it could for example just have crashed a few seconds ago - hence not be active - but in the process of being restarted, with the same slingId. But this case is also covered by the cleanup algorithm as in that case the clusterNodeId would (well, only "very likely") still be in the /idmap and mapped to this same slingId. In which case it woudl not be deleted.
Worst case, there is indeed a race condition between this new cleanup and a Sling instance startup, and it would be able to disturb a startup. In that case, that instance would have to be restarted. 
But I'd consider that likelyhood as "very unlikely" as well ;)

Reply:
Worst-case should be that the existing, active instances get into a TOPOLOGY_CHANGING situation and the restarted, racing instance fails to join the topology as well but the instance is running, so oak leases are exchanged. And killing that problematic instance would be required to get back to a proper topology.

One thing that I'll try to look into is whether each instance shouldn't keep a list of all slingIds that were ever active (while it was alive). That list shouldn't be large, and holding just those slingIds shouldn't cause any memory issue. But having that list and then not cleaning up any slingId of that list would further reduce the likelihood of the described race-condition. Actually I think it would solve it completely.

Reply:
Added code to address this in https://github.com/apache/sling-org-apache-sling-discovery-oak/pull/13/commits/b57b2e1d919f345a59359db179fe7fa8e265b2c2

Reply:
Marking htis conversation as closed, please reopen if you disagree, thx